pkg/terminal/command.go

```diff
@@ -2873,16 +2875,72 @@ func printTracepoint(t *Term, th *api.Thread, bpname string, fn *api.Function, a
 		fmt.Fprintf(t.stdout, "%s ", time.Now().Format(time.RFC3339Nano))
 	}
 
+	var sdepth, wantindex, rootindex int
+	depthPrefix := ""
+	if th.Breakpoint.TraceFollowCalls > 0 {
+		// Trace Follow Calls; stack is required to calculate depth of functions
+		wantindex = -1
+		rootindex = -1
+		if th.BreakpointInfo == nil || th.BreakpointInfo.Stacktrace == nil {
+			return
+		}
+
+		stack := th.BreakpointInfo.Stacktrace
+		for i := range stack {
+			curi := len(stack) - 1 - i
```

Comment:
I agree this is a bit confusing to read through.

So, `rootindex` is the closest frame of the root function (e.g. `foo` in the case of `dlv trace foo --follow-calls`) from the current function, and `wantindex` is the frame of a child call of that function, correct?

E.g. given:

```go
func f1() {
        f2()
}

func f2() {
        f3()
}

func f3() {}
```

Running `dlv trace f1 --follow-calls=3` once we hit `f3` the stack would look something like like:

```
[0] main.main
[1] f1 <-- rootindex
[2] f2
[3] f3 <-- wantindex
```

I think `wantindex` would always equal `len(stack)` right? Because we want to calculate depth from where we are at currently. In that case I don't think we even need to track it with a variable.

In which case couldn't we refactor the current algorithm like so:

```go
for i := len(stack)-1; i > 0; i-- {
        if stack[i].Function.Name() == th.Breakpoint.RootFuncName {
		rootindex = i
                break // no need to continue looping at this point
	}
}
/*
depth we want == 2
len(stack) == 4
rootindex == 1
(len(stack) - 1) - rootindex == 2
*/
sdepth = len(stack) - rootindex - 1
```

However this wouldn't handle recursive function call depth correctly, I think you would want to find `rootindex` closest to the top of the stack, right? Kind of like what you're doing with `wantindex` currently.

As it stands right now I think there might be a bug in your calculation, if I'm understanding correctly. We only assign `rootindex` once (guarded by the `if rootindex == -1` branch) however we assign `wantindex` everytime we see it, because we don't short circuit the loop. So given the same situation above, but if the stack somehow looked like this (this is a contrived example, you can't get into this situation with the above code):

```
[0] main.main
[1] f3 <-- wantindex
[2] f1
[3] f3
[4] f1 <-- rootindex
[5] f3
```
Then in the current algorithm would place `wantindex` at `1` and the calculation would be:

```
wantindex == 1
rootindex == 4
4 - 1 + 1 == 4
```

Which would be incorrect. Again, please correct me if I am misunderstanding the current algorithm.

I think it may be worth extracting this code into it's own function and unit testing it to ensure correctness in different scenarios.

Reply:
I've been thinking about this a bit more and there are two possible interpretation to this feature:

1. the "static" interpretation: tracepoints are set on all functions reachable from the root function up to the specified depth. These tracepoints then behave like normal tracepoints, so, for example, if we set a tracepoint on `f` with `--follow-calls=1` and `f` recursively calls itself to a depth of 10 we print all ten recursive calls
2. the "dynamic" interpretation: all calls made from the root function up to the specified depth are traced. If we set a tracepoint on `f` with `--follow-calls=1` and `f` recursively calls itself to a depth of 10 we only print the first call.

I thought we were going for the "static" interpretation (that's what it did originally and the "dynamic" interpretation is pretty hard to do), but the current implementation is a mix of both. It shouldn't be.

Reply:
I updated my comment above because I had a few things reversed which may have made my thought a bit more difficult to understand. Also, the issue with recursion isn't an issue at all as I thought about it more.

The more I think about this, I believe this would always be correct:

```go
/*
Given the following stack:

[5] runtime.goexit
[4] runtime.main
[3] main.main
[2] f1 <-- rootindex
[1] f2
[0] f3 <-- wantindex
*/

// We do not need to visit stack[0] (current frame) because that will always be `wantindex`, the child function
// which triggered the breakpoint.
for i := len(stack)-1; i > 0; i-- {
        if stack[i].Function.Name() == th.Breakpoint.RootFuncName {
		sdepth = i
                break // no need to continue looping at this point
	}
}
```

In other words, we don't need to do any arithmetic at all. Since we're going from the top of the stack down towards the bottom (current frame), the iterator value of the rootindex will always be equal to the depth.

We could rework the algorithm to go from the bottom of the stack (current frame) up, and use the iterator as the depth:

```go
for i := range stack {
    if stack[i].Function.Name() == th.Breakpoint.RootFuncName {
        sdepth = i // always overwrite sdepth whenever we see the root function
    }
}
```

However it is less efficient because we would have to traverse the entire stack every time, whereas with the current approach of going top down we can short circuit and break as soon as we see the root function appear.

Reply:
the new version implements the short circuit idea